### PR TITLE
HIVE-26062: Make sure that running 4.0.0-alpha-2 above 4.0.0-alpha-1 …

### PR DESCRIPTION
**Pull request by dengzhhu653**
…HMS DB reports the correct error

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreSchemaInfo.java [hunk 1]
```diff
@@ -205,19 +206,33 @@ public boolean isVersionCompatible(String hiveVersion, String dbVersion) {
...
+  private int compareVersion(String dbVerPart, String hiveVerPart) {
+    if (dbVerPart.equals(hiveVerPart)) {
+      return 0;
+    }
+    boolean isDbVerNum = NumberUtils.isNumber(dbVerPart);
+    boolean isHiveVerNum = NumberUtils.isNumber(hiveVerPart);
+    if (isDbVerNum && isHiveVerNum) {
+      return Integer.parseInt(dbVerPart) - Integer.parseInt(hiveVerPart);
+    } else if (!isDbVerNum && !isHiveVerNum) {
+      return dbVerPart.compareTo(hiveVerPart);
```
[the remark this answers is not in the document]
Comment (pvary) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreSchemaInfo.java:232]: Makes sense

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreSchemaInfo.java [hunk 2]
```diff
@@ -205,19 +206,32 @@ public boolean isVersionCompatible(String hiveVersion, String dbVersion) {
...
-    for (int i = 0; i < dbVerParts.length; i++) {
-      int dbVerPart = Integer.parseInt(dbVerParts[i]);
-      int hiveVerPart = Integer.parseInt(hiveVerParts[i]);
-      if (dbVerPart > hiveVerPart) {
-        return true;
-      } else if (dbVerPart < hiveVerPart) {
-        return false;
-      } else {
-        continue; // compare next part
+    int i = 0;
```
Comment (pvary) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreSchemaInfo.java:209]: This makes me nervous 😄 
Why do we have to move around the declaration of `i`?

Maybe the code would be easier to read this way:
```
hiveVerParts = hiveVersion.split("\\.|-");
dbVerParts = dbVersion.split("\\.|-");
for (int i; i < Math.min(hiveVerParts.length, dbVerParts.length); i++) {
  int compare = compareVersion(dbVerParts[i], hiveVerParts[i]);
  if (compare != 0) {
    return compare > 0;
  }
}

// This is always false, as we already checked that hiveVersion and dbVersions are different 
return hiveVerParts.length == dbVerParts.length;
```

Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreSchemaInfo.java:209]: If `i` reaches the end of `dbVerParts`, means that the hive version is behind the db version, which supposed to be compatiable. For example,
dbVerParts is `4`, `0`, `0`
hiveVerParts is `4`, `0`, `0`, `alpha`, `1`

the `hiveVerParts.length == dbVerParts.length;` returns false in this case.

Comment (pvary) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreSchemaInfo.java:209]: Then this should be always `true`?
Or
`hiveVerParts.length > dbVerParts.length`?

Comment (dengzhhu653) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreSchemaInfo.java:209]: `hiveVerParts.length > dbVerParts.length` makes sense, I will fix that. Thank you for your suggestion!

**Pull request merged**